File: src/lib/modelCapabilities.ts
/**
 * Explicit model-input capability model.
 *
 * `supportsVision` remains the persisted legacy image flag. New callers may
 * provide this profile for modality-specific decisions without changing old
 * provider settings that only carry `supportsVision`.
 */
export type CapabilitySupport = "supported" | "unsupported" | "unknown";

export type AttachmentKind =
  | "image"
  | "video"
  | "audio"
  | "document"
  | "file"
  | "directory";

export interface ModelCapabilities {
  /** Direct image pixels / image content blocks. */
  imageInput?: CapabilitySupport;
  /** Direct video input, not a path reference. */
  videoInput?: CapabilitySupport;
  /** Direct audio input, not a path reference. */
  audioInput?: CapabilitySupport;
  /** Native document content input. */
  documentInput?: CapabilitySupport;
  /** Local file/directory references such as `@/path`. */
  fileReference?: CapabilitySupport;
  /** Can preserve an unfinished tool call/result causal chain. */
  toolCausalContinuation?: CapabilitySupport;
  /** Can continue facts that are bound to the originating provider. */
  providerBoundContinuation?: CapabilitySupport;
}

const capabilityValue = (
  value: CapabilitySupport | undefined,
): CapabilitySupport => value ?? "unknown";

/** Convert the persisted `supportsVision` flag into an explicit image capability. */
export function capabilitiesFromLegacyVision(
  supportsVision: boolean | undefined,
): ModelCapabilities {
  return {
    imageInput:
      supportsVision === undefined
        ? "unknown"
        : supportsVision
          ? "supported"
          : "unsupported",
  };
}

/** Explicit profile fields take precedence; legacy image capability fills gaps. */
export function mergeModelCapabilities(
  explicit: ModelCapabilities | undefined,
  supportsVision: boolean | undefined,
): ModelCapabilities {
  return {
    ...capabilitiesFromLegacyVision(supportsVision),
    ...explicit,
  };
}

export function capabilityForAttachmentKind(
  capabilities: ModelCapabilities | undefined,
  kind: AttachmentKind,
): CapabilitySupport {
  const c = capabilities ?? {};
  switch (kind) {
    case "image":
      return capabilityValue(c.imageInput);
    case "video":
      return capabilityValue(c.videoInput);
    case "audio":
      return capabilityValue(c.audioInput);
    case "document":
      return capabilityValue(c.documentInput);
    case "file":
    case "directory":
      return capabilityValue(c.fileReference);
  }
}

/** A legacy text-only target has explicitly rejected direct image input. */
export function isTextOnlyTarget(
  capabilities: ModelCapabilities | undefined,
): boolean {
  return capabilityValue(capabilities?.imageInput) === "unsupported";
}

File: src-tauri/src/model_capabilities.rs
//! Explicit input/continuation capabilities for a model.
//!
//! Legacy provider settings only contain `supports_vision`; their value is
//! mapped into `image_input` at the boundary so existing config remains valid.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CapabilitySupport {
    Supported,
    Unsupported,
    Unknown,
}

impl Default for CapabilitySupport {
    fn default() -> Self {
        Self::Unknown
    }
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelCapabilities {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub image_input: Option<CapabilitySupport>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub video_input: Option<CapabilitySupport>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub audio_input: Option<CapabilitySupport>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub document_input: Option<CapabilitySupport>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub file_reference: Option<CapabilitySupport>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_causal_continuation: Option<CapabilitySupport>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provider_bound_continuation: Option<CapabilitySupport>,
}

impl ModelCapabilities {
    pub fn image_support(&self) -> CapabilitySupport {
        self.image_input.unwrap_or_default()
    }

    pub fn with_legacy_vision(mut self, supports_vision: Option<bool>) -> Self {
        if self.image_input.is_none() {
            self.image_input = supports_vision.map(|supported| {
                if supported {
                    CapabilitySupport::Supported
                } else {
                    CapabilitySupport::Unsupported
                }
            });
        }
        self
    }
}
